File: leetcode/easy/1051.Height Checker/python/heightChecker.py
"""Height Checker"""
def height_checker(heights):
    #edge cases
    sorted_heights = sorted(heights)
    if sorted_heights == heights:
        return 0
    #normal cases
    else:
        wrong_positions = 0
        for i in range(len(heights)):
            if heights[i] != sorted_heights[i]:
                wrong_positions += 1
        return wrong_positions
                    

heights =  [1,1,4,2,1,3]
heights1 = [5,1,2,3,4]
expected = [1,1,1,2,3,4]

print(height_checker(heights))

'''
1.
input: integer array
output: integer number which represents all the indexes that don't match
heights:  [1,1,4,2,1,3]
expected: [1,1,1,2,3,4]
output: 3

2.
edge cases: increasing order return 0
key: it is not neccessary to sort the array only compare it with 
the sorted array and keep track of all the indexes that don't match
with the sorted array

3. 
sort_heights = sort(heights)
conditional statement if sort_heights == to heights if so return 0
else
    for loop in order to iterate each element with the sorted heights
        conditional statement to know if n element is different to its contrapart if so
            then add 1 to the count variable
return the count variable
    
4. coding

'''
def height_checker(heights):
    unorder_heights = heights[:]
    heights.sort()
    count_wrong_student_place = 0
    for i in range(len(heights)):
        if heights[i] != unorder_heights[i]:
            count_wrong_student_place += 1
    return count_wrong_student_place

# Testing 
data_test_0 = [1,1,4,2,1,3]
data_test_1 = [5,1,2,3,4]
data_test_2 = [1,2,3,4,5]
print(height_checker(data_test_2))

'''Complexity Analysis:
 * Time Complexity: O(NlogN)
 * Space Complexity: O(N)'''

File: leetcode/easy/1051.Height Checker/js/heightChecker.js
"use strict";

/**
 * @param {number[]} heights
 * @return {number} amountIndexes
 */

// 1. burble sort new sortHeights
// 1.1. isItSorted = false
// 1.2 while isItSorted equal false ...
// isItSorted = true
// for i equal 0, while i < heights length...
// ... if heights[i] > heights[i + 1] ... isItSorted = false
// temp = heights[i] heights[i] = heights[i + 1], heights[i +1] = temp
// 2. countIndexes = 0
// 3. for heights
// 3.1 if heights[i] != sortedHeights[i]... countIndexes += 1
// 4. return countIndexes
var heightChecker = function(heights) {
    let sortedHeights = heights.slice();
    let isItSorted = false;
    console.log(sortedHeights);
    while (isItSorted === false) {
        isItSorted = true;
        for (let i = 0; i < sortedHeights.length - 1; i++) {
            if (sortedHeights[i] > sortedHeights[i + 1]) {
                isItSorted = false;
                let temp = sortedHeights[i];
                sortedHeights[i] = sortedHeights[i + 1];
                sortedHeights[i + 1] = temp;
            };
        };
    };
    let countIndexes = 0;
    for (let i = 0; i < sortedHeights.length; i++) {
        if (sortedHeights[i] !== heights[i]) {
            countIndexes++;
        };
    };
    console.log(sortedHeights);
    console.log(heights);
    return countIndexes;
};

let heights = [1,1,4,2,1,3];
console.log(heightChecker(heights));

var heightChecker = function(heights) {
    let sortedHeights = heights.slice().sort((a, b) => a - b);
    let countWrongStudentPlaces = 0;
    for (let i = 0; i < heights.length; i++) {
        if (heights[i] !== sortedHeights[i]) {
            countWrongStudentPlaces++;
        };
    };
    return countWrongStudentPlaces;
};

// Testing 
let dataTest1 = [5,1,2,3,4];
console.log(heightChecker(dataTest1));

/**Complexity Analysis:
 * Time Complexity: O(NlogN)
 * Space Complexity: O(N)
 */